### PR TITLE
fix pull image error from multiple ACRs using azure managed identity

### LINKED ISSUE #92326: Pulling Azure Container Registry image using Managed Service Identity may fail
**What happened**:
When Managed Service Identity (MSI) is used to pull images from Azure Container Registry (ACR), it may fail with `Failed to pull image "<redacted>.azurecr.io/<redacted>": rpc error: code = Unknown desc = Error response from daemon: Get https://<redacted>.azurecr.io/v2/<redacted>: unauthorized: authentication required` even though the MSI has reader access to the ACR. It will recover in 1-2 minutes.

**What you expected to happen**:
It should not have any unauthorized error.

**How to reproduce it (as minimally and precisely as possible)**:
1. run a k8s deployment using public docker image
2. immediately run a k8s deployment using ACR image
3. observe the error in get events
```
66s         Normal    Scheduled           Pod          Successfully assigned default/nginx-deployment-7fd6966748-txsdm to aks-agentpool-17585647-vmss000001
65s         Normal    Pulling             Pod          Pulling image "nginx:1.14.2"
61s         Normal    Pulled              Pod          Successfully pulled image "nginx:1.14.2"
58s         Normal    Created             Pod          Created container nginx
58s         Normal    Started             Pod          Started container nginx
66s         Normal    Scheduled           Pod          Successfully assigned default/nginx-deployment-7fd6966748-xmk6h to aks-agentpool-17585647-vmss000000
65s         Normal    Pulling             Pod          Pulling image "nginx:1.14.2"
61s         Normal    Pulled              Pod          Successfully pulled image "nginx:1.14.2"
58s         Normal    Created             Pod          Created container nginx
58s         Normal    Started             Pod          Started container nginx
66s         Normal    SuccessfulCreate    ReplicaSet   Created pod: nginx-deployment-7fd6966748-xmk6h
66s         Normal    SuccessfulCreate    ReplicaSet   Created pod: nginx-deployment-7fd6966748-txsdm
66s         Normal    SuccessfulCreate    ReplicaSet   Created pod: nginx-deployment-7fd6966748-bxb62
66s         Normal    ScalingReplicaSet   Deployment   Scaled up replica set nginx-deployment-7fd6966748 to 3
46s         Normal    Scheduled           Pod          Successfully assigned default/ubuntu-56b4744f8b-6mcpk to aks-agentpool-17585647-vmss000000
4s          Normal    Pulling             Pod          Pulling image "<redacted>.azurecr.io/<redacted>"
30s         Warning   Failed              Pod          Failed to pull image "<redacted>.azurecr.io/<redacted>": rpc error: code = Unknown desc = Error response from daemon: Get https://<redacted>.azurecr.io/v2/<redacted>: unauthorized: authentication required
30s         Warning   Failed              Pod          Error: ErrImagePull
16s         Normal    BackOff             Pod          Back-off pulling image "<redacted>.azurecr.io/<redacted>"
16s         Warning   Failed              Pod          Error: ImagePullBackOff
```

Kubelet log:
```
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.421665   29650 provider.go:124] Refreshing cache for provider: *credentialprovider.defaultDockerConfigProvider
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422075   29650 config.go:137] looking for config.json at /var/lib/kubelet/config.json
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422119   29650 config.go:137] looking for config.json at /config.json
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422130   29650 config.go:137] looking for config.json at /.docker/config.json
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422137   29650 config.go:137] looking for config.json at /.docker/config.json
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422148   29650 config.go:107] looking for .dockercfg at /var/lib/kubelet/.dockercfg
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422177   29650 config.go:107] looking for .dockercfg at /.dockercfg
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422186   29650 config.go:107] looking for .dockercfg at /.dockercfg
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422194   29650 config.go:107] looking for .dockercfg at /.dockercfg
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422202   29650 provider.go:94] Unable to parse Docker config file: couldn't find valid .dockercfg after checking in [/var/lib/kubelet   /]
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422210   29650 provider.go:124] Refreshing cache for provider: *azure.acrProvider
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422232   29650 azure_credentials.go:189] try to provide secret for image docker.io/library/nginx
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422244   29650 azure_credentials.go:194] image(docker.io/library/nginx) is not from ACR, skip MSI authentication
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.422262   29650 kuberuntime_image.go:47] Pulling image "nginx:1.14.2" without credentials
Jun 19 22:36:33 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:33.421975   29650 event.go:258] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"nginx-deployment-7fd6966748-xmk6h", UID:"128464dd-b9d3-4e54-a3f9-77d7af6e3be2", APIVersion:"v1", ResourceVersion:"158317", FieldPath:"spec.containers{nginx}"}): type: 'Normal' reason: 'Pulling' Pulling image "nginx:1.14.2"
Jun 19 22:36:37 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:37.096673   29650 kube_docker_client.go:345] Stop pulling image "nginx:1.14.2": "Status: Downloaded newer image for nginx:1.14.2"

Jun 19 22:36:54 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:54.067201   29650 event.go:258] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"ubuntu-56b4744f8b-6mcpk", UID:"d9d46d4b-7d40-4c10-82c4-07a36a788d7b", APIVersion:"v1", ResourceVersion:"158402", FieldPath:"spec.containers{ubuntu}"}): type: 'Normal' reason: 'Pulling' Pulling image "<redacted>.azurecr.io/<redacted>"
Jun 19 22:36:54 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:54.175261   29650 kubelet.go:1975] SyncLoop (housekeeping)
Jun 19 22:36:54 aks-agentpool-17585647-vmss000000 kubelet[29650]: E0619 22:36:54.310523   29650 remote_image.go:113] PullImage "<redacted>.azurecr.io/<redacted>" from image service failed: rpc error: code = Unknown desc = Error response from daemon: Get https://<redacted>.azurecr.io/v2/fundamental/test/manifests/v0.0.12: unauthorized: authentication required
Jun 19 22:36:54 aks-agentpool-17585647-vmss000000 kubelet[29650]: E0619 22:36:54.310655   29650 kuberuntime_manager.go:779] container start failed: ErrImagePull: rpc error: code = Unknown desc = Error response from daemon: Get https://<redacted>.azurecr.io/v2/fundamental/test/manifests/v0.0.12: unauthorized: authentication required
Jun 19 22:36:54 aks-agentpool-17585647-vmss000000 kubelet[29650]: E0619 22:36:54.310682   29650 pod_workers.go:190] Error syncing pod d9d46d4b-7d40-4c10-82c4-07a36a788d7b ("ubuntu-56b4744f8b-6mcpk_default(d9d46d4b-7d40-4c10-82c4-07a36a788d7b)"), skipping: failed to "StartContainer" for "ubuntu" with ErrImagePull: "rpc error: code = Unknown desc = Error response from daemon: Get https://<redacted>.azurecr.io/v2/fundamental/test/manifests/v0.0.12: unauthorized: authentication required"
Jun 19 22:36:54 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:54.310974   29650 event.go:258] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"ubuntu-56b4744f8b-6mcpk", UID:"d9d46d4b-7d40-4c10-82c4-07a36a788d7b", APIVersion:"v1", ResourceVersion:"158402", FieldPath:"spec.containers{ubuntu}"}): type: 'Warning' reason: 'Failed' Failed to pull image "<redacted>.azurecr.io/<redacted>": rpc error: code = Unknown desc = Error response from daemon: Get https://<redacted>.azurecr.io/v2/fundamental/test/manifests/v0.0.12: unauthorized: authentication required
Jun 19 22:36:54 aks-agentpool-17585647-vmss000000 kubelet[29650]: I0619 22:36:54.311026   29650 event.go:258] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"ubuntu-56b4744f8b-6mcpk", UID:"d9d46d4b-7d40-4c10-82c4-07a36a788d7b", APIVersion:"v1", ResourceVersion:"158402", FieldPath:"spec.containers{ubuntu}"}): type: 'Warning' reason: 'Failed' Error: ErrImagePull
```
**Anything else we need to know?**:
When MSI is configured in kubelet and a non-ACR docker image is pulled, the code [here](https://github.com/kubernetes/kubernetes/blob/v1.15.10/pkg/credentialprovider/azure/azure_credentials.go#L194) will skip the MSI authentication flow and [return an empty docker registry cfg](https://github.com/kubernetes/kubernetes/blob/v1.15.10/pkg/credentialprovider/azure/azure_credentials.go#L213) assuming it's anonymous ACR auth. The resulting docker registry cfg is [cached in the credential provider](https://github.com/kubernetes/kubernetes/blob/v1.15.10/pkg/credentialprovider/provider.go#L120). So any image pull from ACR will use the staled docker registry cfg and got unauthorized authentication until the cache is refreshed

**Environment**:
- Kubernetes version (use `kubectl version`): 1.15.10
- Cloud provider or hardware configuration: AKS with MSI

### PR DESCRIPTION
**Pull request by andyzhangx**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
fix pull image error from multiple ACRs using azure managed identity

This PR continues with PR(https://github.com/kubernetes/kubernetes/pull/92330), fixed the pull image error with azure managed identity.

This PR is using per-registry cache (also with cache expiration), pls note that per-registry cache is only for managed identity, and for service principal(account name & password), it's not necessary since password is supposed to be the same in whole kubelet life time.

**Which issue(s) this PR fixes**:

Fixes #92326

**Special notes for your reviewer**:
 - impact of this issue, you may hit this issue when pulling images from multiple different ACRs, and after a few more retries, it should work:
```
Events:
  Type     Reason     Age                 From                                        Message
  ----     ------     ----                ----                                        -------
  Normal   Scheduled  104s                default-scheduler                           Successfully assigned default/nginx2 to aks-nodepool1-43715599-vmss000001
  Warning  Failed     55s (x3 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Failed to pull image "andyacr3.azurecr.io/nginx:v1": rpc error: code = Unknown desc = Error response from daemon: Get https://andyacr3.azurecr.io/v2/nginx/manifests/v1: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information.
  Warning  Failed     55s (x3 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Error: ErrImagePull
  Normal   BackOff    27s (x4 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Back-off pulling image "andyacr3.azurecr.io/nginx:v1"
  Warning  Failed     27s (x4 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Error: ImagePullBackOff
  Normal   Pulling    12s (x4 over 103s)  kubelet, aks-nodepool1-43715599-vmss000001  Pulling image "andyacr3.azurecr.io/nginx:v1"
  Normal   Pulled     12s                 kubelet, aks-nodepool1-43715599-vmss000001  Successfully pulled image "andyacr3.azurecr.io/nginx:v1" in 383.922005ms
  Normal   Created    12s                 kubelet, aks-nodepool1-43715599-vmss000001  Created container nginx
  Normal   Started    12s                 kubelet, aks-nodepool1-43715599-vmss000001  Started container nginx
```

With this fix, there is no authentication error when pulling images from two ACRs using managed identity:
```
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  14s   default-scheduler                           Successfully assigned default/nginx to aks-nodepool1-43715599-vmss000000
  Normal  Pulling    13s   kubelet, aks-nodepool1-43715599-vmss000000  Pulling image "andyacr2.azurecr.io/nginx:v1"
  Normal  Pulled     6s    kubelet, aks-nodepool1-43715599-vmss000000  Successfully pulled image "andyacr2.azurecr.io/nginx:v1" in 7.513394748s
  Normal  Created    2s    kubelet, aks-nodepool1-43715599-vmss000000  Created container nginx
  Normal  Started    2s    kubelet, aks-nodepool1-43715599-vmss000000  Started container nginx

Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  13s   default-scheduler                           Successfully assigned default/nginx2 to aks-nodepool1-43715599-vmss000000
  Normal  Pulling    13s   kubelet, aks-nodepool1-43715599-vmss000000  Pulling image "andyacr3.azurecr.io/nginx:v1"
  Normal  Pulled     2s    kubelet, aks-nodepool1-43715599-vmss000000  Successfully pulled image "andyacr3.azurecr.io/nginx:v1" in 10.224016565s
  Normal  Created    2s    kubelet, aks-nodepool1-43715599-vmss000000  Created container nginx
  Normal  Started    2s    kubelet, aks-nodepool1-43715599-vmss000000  Started container nginx
```

**Does this PR introduce a user-facing change?**:

```release-note
fix pull image error from multiple ACRs using azure managed identity
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
fix pull image error from multiple ACRs using azure managed identity
```

/priority important-soon
/sig cloud-provider
/area provider/azure
/triage accepted
/assign @feiskyer

### CODE REVIEW

**Comment by andyzhangx** [comment]
/hold cancel
test pass: With this fix, there is no authentication error when pulling images from two ACRs using managed identity

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 1]
```diff
@@ -186,6 +191,30 @@ func (a *acrProvider) Enabled() bool {
...
+			loginServer := a.parseACRLoginServerFromImage(image)
+			if loginServer == "" {
+				return a.cacheDockerConfig
+			}
+
+			// check whether loginServer is in cache
+			for k := range a.cacheDockerConfig {
+				if strings.EqualFold(loginServer, k) {
+					return a.cacheDockerConfig
+				}
```
Comment (liggitt) [inline pkg/credentialprovider/azure/azure_credentials.go:209]: We shouldn't add logic to this plugin that could not be expressed within the API of the externalized credential provider plugins, as that will prevent moving this credential provider out of tree.

Can you give a concrete example of what the `image` string will be, and what the `loginServer` string will be?

It seems like this plugin is fetching credentials based on the image and wants to cache them for any image using the same registry. Is that the case?

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:209]: @liggitt below is an example, with Azure Managed Identity (no password), the credential is based on `loginServer`, and different `loginServer` has different credentials.
 - `image`: `andyacr3.azurecr.io/nginx:v1`
 - `loginServer`: `andyacr3.azurecr.io`

AWS also uses same implementation today:
https://github.com/kubernetes/kubernetes/blob/4261200724cd0541b4ff02cc56be866b98142823/pkg/credentialprovider/aws/aws_credentials.go#L85-L98
I think externalized credential provider plugin should also supports such logic.

Anyway, this is now a critical bug in Azure side, we need to fix it ASAP, thanks for understanding, and will check how to bring this logic to externalized credential provider plugin later.

Comment (liggitt) [reply pkg/credentialprovider/azure/azure_credentials.go:209]: Does the login server always match the registry portion of the image?

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:209]: depends on image itself, if it's Azure Container Registry image, e.g. `andyacr3.azurecr.io/nginx:v1`, then it would always match since we only cache Azure Container Registry login server.

**Comment by andrewsykim** [comment]
> We shouldn't add logic to this plugin that could not be expressed within the API of the externalized credential provider plugins, as that will prevent moving this credential provider out of tree.

I agree with @liggitt that we shouldn't add to the existing credentials providers, especially if the externalized API will handle this exact use-case (cachinig credentials on a per-registry basis).

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 2]
```diff
@@ -186,6 +191,30 @@ func (a *acrProvider) Enabled() bool {
...
 func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// If the cache hasn't expired, return our cache
+	if time.Now().Before(a.expiration) {
+		if a.config != nil && a.config.UseManagedIdentityExtension {
+			loginServer := a.parseACRLoginServerFromImage(image)
+			if loginServer == "" {
+				return a.cacheDockerConfig
```
Comment (andrewsykim) [inline pkg/credentialprovider/azure/azure_credentials.go:202]: This sort of behavior is not something the externalized API will cover, if we can't determine the registry (login server), we should log an error and return an empty `credntialprovider.DockerConfig`

Comment (andrewsykim) [reply pkg/credentialprovider/azure/azure_credentials.go:202]: This would also result in the ACR provider returning cached docker config for non-ACR repos?

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 3]
```diff
@@ -186,6 +191,30 @@ func (a *acrProvider) Enabled() bool {
...
+			}
+
+			// check whether loginServer is in cache
+			for k := range a.cacheDockerConfig {
+				if strings.EqualFold(loginServer, k) {
+					return a.cacheDockerConfig
+				}
+			}
+			// continue with get ARM token next since loginServer not found in cache
```
Comment (andrewsykim) [inline pkg/credentialprovider/azure/azure_credentials.go:211]: Since we removed CachingDockerConfig, the code path to fetch ARM token would run for any non-ACR repo where the loginServer is never a match. Is that the case here?

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:211]: that's not correct, see Line 229: `klog.V(4).Infof("image(%s) is not from ACR, skip MSI authentication", image)`, if it's non-ACR repo, it won't fetch ARM token

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:211]: anyway, return empty DockerConfig for non-ACR repo looks reasonable, will fix it soon, thanks.

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:211]: fixed in line 202, PTAL.

File: pkg/credentialprovider/azure/azure_credentials.go (hunk 2 above)
Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go]: fixed in line 203, PTAL.

**Comment by andyzhangx** [comment]
I refactored the logic in `Provide` func in Azure provider a little, pulling 3 images from two ACRs and dockerhub successfully, PTAL. Thanks.
```
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  25s   default-scheduler                           Successfully assigned default/nginx to aks-nodepool1-43715599-vmss000002
  Normal  Pulling    24s   kubelet, aks-nodepool1-43715599-vmss000002  Pulling image "andyacr2.azurecr.io/nginx:v1"
  Normal  Pulled     16s   kubelet, aks-nodepool1-43715599-vmss000002  Successfully pulled image "andyacr2.azurecr.io/nginx:v1" in 8.3201804s
  Normal  Created    12s   kubelet, aks-nodepool1-43715599-vmss000002  Created container nginx
  Normal  Started    11s   kubelet, aks-nodepool1-43715599-vmss000002  Started container nginx
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  25s   default-scheduler                           Successfully assigned default/nginx to aks-nodepool1-43715599-vmss000002
  Normal  Pulling    24s   kubelet, aks-nodepool1-43715599-vmss000002  Pulling image "andyacr2.azurecr.io/nginx:v1"
  Normal  Pulled     16s   kubelet, aks-nodepool1-43715599-vmss000002  Successfully pulled image "andyacr2.azurecr.io/nginx:v1" in 8.3201804s
  Normal  Created    12s   kubelet, aks-nodepool1-43715599-vmss000002  Created container nginx
  Normal  Started    11s   kubelet, aks-nodepool1-43715599-vmss000002  Started container nginx
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  24s   default-scheduler                           Successfully assigned default/nginx3 to aks-nodepool1-43715599-vmss000002
  Normal  Pulling    23s   kubelet, aks-nodepool1-43715599-vmss000002  Pulling image "nginx"
  Normal  Pulled     12s   kubelet, aks-nodepool1-43715599-vmss000002  Successfully pulled image "nginx" in 11.0415041s
  Normal  Created    12s   kubelet, aks-nodepool1-43715599-vmss000002  Created container nginx
  Normal  Started    12s   kubelet, aks-nodepool1-43715599-vmss000002  Started container nginx
```

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 4]
```diff
@@ -186,24 +191,45 @@ func (a *acrProvider) Enabled() bool {
...
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// If the cache hasn't expired, return our cache
+	if time.Now().Before(a.expiration) {
+		if a.config != nil && a.config.UseManagedIdentityExtension {
+			// check whether loginServer is in cache
+			for k := range a.cacheDockerConfig {
+				if strings.EqualFold(loginServer, k) {
+					return a.cacheDockerConfig
```
Comment (andrewsykim) [inline pkg/credentialprovider/azure/azure_credentials.go:210]: I still don't think this does what we want. If there are two subsequent calls to `Provide` with different login servers `foo1.azurecr.io` and `foo2.azurecr.io`, the second call will update `cacheDockerConfig` and invalidate caching for foo1.

Comment (liggitt) [reply pkg/credentialprovider/azure/azure_credentials.go:210]: Agreed. Since this provider is caching on the registry level, its implementation should reflect that. This is essentially the request made in https://github.com/kubernetes/kubernetes/pull/92818#issuecomment-656261189 (to make the cache coherent with the significant portion of the image input).

It looks like the registry is the significant portion of the image input that affects the computed results, so the cache should be keyed by the registry when UseManagedIdentityExtension is enabled:

```
// cache credentials globally
cacheKey := ""
if a.config != nil && a.config.UseManagedIdentityExtension {
  // cache credentials at the registry level
  cacheKey = strings.Split(image, "/")[0]
}

cachedCredentials, found := cache.getUnexpiredCredentials(cacheKey)
if found {
  return cachedCredentials
}

... get credentials and store in cache under cacheKey ...
```

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:209]: @andrewsykim thanks, I have fixed it, see L213-L215

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:209]: @andrewsykim @liggitt could you take a look again? thanks
here `loginServer` is actually `cacheKey`, it's actually identical though with a little different implementation.

Comment (liggitt) [reply pkg/credentialprovider/azure/azure_credentials.go:209]: > here `loginServer` is actually `cacheKey`, it's actually identical though with a little different implementation.
> `cacheDockerConfig[loginServer] = credential`

It would be helpful to make it clear that this logic can fit within the externalized credential provider API to compute the cache key the same way the provider will (taking the registry portion of the image, and not adding in case-insensitive comparisons)

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 5]
```diff
@@ -186,24 +191,46 @@ func (a *acrProvider) Enabled() bool {
...
+				}
+			}
+			// continue with get ARM token next since loginServer not found in cache
+			if cred, err := getACRDockerEntryFromARMToken(a, loginServer); err == nil {
+				a.cacheDockerConfig[loginServer] = *cred
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:225]: fixed.

**Comment by andyzhangx** [comment]
@liggitt I changed the logic a little per comments, removed the `case-insensitive comparisons`, it does not matter if it's case-sensitive, try get one more token is ok.

File: pkg/credentialprovider/azure/azure_credentials.go (hunk 4 above)
Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go]: I have removed the case-insensitive comparisons, should fit within externalized credential provider API now. Let's merge this PR to get in-tree code work first, thanks.  @liggitt

File: pkg/credentialprovider/azure/azure_credentials.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/credentialprovider/azure/azure_credentials.go:225]: this is still modifying a value that was previously returned to callers (line 273-274 assign the returned value to cacheDockerConfig), so this is still a data race

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 6]
```diff
@@ -237,10 +267,10 @@ func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
 				cfg[customAcrSuffix] = *cred
 			}
 		}
-
-		// add ACR anonymous repo support: use empty username and password for anonymous access
-		cfg["*.azurecr.*"] = defaultConfigEntry
 	}
+
+	a.expiration = time.Now().Add(a.lifetime)
```
Comment (liggitt) [inline pkg/credentialprovider/azure/azure_credentials.go:272]: This is lengthening the expiration time on existing credentials copied from any previous cacheDockerConfig (line 203). Do the acquired tokens expire? if so, that means requests for images in new registries can extend the cache expiration even as individual items in cacheDockerConfig are no longer valid.

Comment (liggitt) [reply pkg/credentialprovider/azure/azure_credentials.go:272]: Having a single cache expiration and cached credential represent data for distinct registries and with distinct expirations is very hard to follow and error prone. Keying the cache by the registry of the incoming image and managing cached credentials+expirations per-registry would make more sense.

Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go]: thanks @liggitt I have updated this PR using per-registry cache (also with cache expiration), pls note that per-registry cache is only for managed identity, and for service principal(account name & password), it's not necessary since password is supposed to be the same in whole kubelet life time.

**Review by feiskyer — COMMENTED** [review]
/lgtm
/hold (in case @liggitt may want to review again)

**Comment by feiskyer** [comment]
/retest
(though azuredisk CSI test failures are not related to this PR's changes)

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 7]
```diff
@@ -237,10 +296,15 @@ func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
...
 
-		// add ACR anonymous repo support: use empty username and password for anonymous access
-		cfg["*.azurecr.*"] = defaultConfigEntry
+	// add ACR anonymous repo support: use empty username and password for anonymous access
+	defaultConfigEntry := credentialprovider.DockerConfigEntry{
+		Username: "",
+		Password: "",
+		Email:    dummyRegistryEmail,
 	}
+	cfg["*.azurecr.*"] = defaultConfigEntry
```
Comment (liggitt) [inline pkg/credentialprovider/azure/azure_credentials.go:307]: in the case where we're returning a cache entry, this is modifying the shared cache object that was previously returned and can cause a data race.

Suggest making an `addAnonymousCredentials(cfg)` method and calling it from `getFromACR` when we're constructing a new cached credential set, and from the else block in Provide() when we're not using UseManagedIdentityExtension. That ensures we never modify a config after we've returned it from Provide()

File: pkg/credentialprovider/azure/azure_credentials.go (hunk 6 above)
Comment (liggitt) [reply pkg/credentialprovider/azure/azure_credentials.go]: that looks much better, thanks. one last comment on how we're adding in anonymous credentials to avoid a data race

File: pkg/credentialprovider/azure/azure_credentials.go (hunk 7 above)
Comment (andyzhangx) [reply pkg/credentialprovider/azure/azure_credentials.go:307]: hi @liggitt there is no shared cache object anymore, `cfg` is a newly constructed object every time in `Provide()` func, see line 253, I have removed the `cacheDockerConfig` field in this PR.

Comment (liggitt) [reply pkg/credentialprovider/azure/azure_credentials.go:307]: ah, I see, thanks.

**Review by liggitt — COMMENTED** [review]
/lgtm
/approve

**Comment by andyzhangx** [comment]
/hold cancel
@liggitt  thanks for the review. Really appreciate that!

**Comment by porrascarlos802018** [comment]
Please update us on this
https://github.com/kubernetes/kubernetes/pull/92818
says will be addressed on   #96355  however , this one at the end says on hold. 

Please advise on ETA and actions required to be executed on existing clusters with this MSI problem.

**Comment by andyzhangx** [comment]
> Please update us on this
> #92818
> says will be addressed on #96355 however , this one at the end says on hold.
> 
> Please advise on ETA and actions required to be executed on existing clusters with this MSI problem.

@porrascarlos802018 this "pull image auth error issue from **multiple**(>1) ACRs using azure managed identity" has been merged in 1.17.14, 1.18.12, 1.19.4, for existing clusters with MSI, wait for a few more tries or using single ACR could workaround
